File: app/src/main/java/com/hjc/reader/ui/menu/adapter/ArticleAdapter.java
package com.hjc.reader.ui.menu.adapter;

import androidx.databinding.DataBindingUtil;

import com.blankj.utilcode.util.StringUtils;
import com.chad.library.adapter.base.BaseDelegateMultiAdapter;
import com.chad.library.adapter.base.delegate.BaseMultiTypeDelegate;
import com.chad.library.adapter.base.viewholder.BaseViewHolder;
import com.hjc.reader.R;
import com.hjc.reader.bean.response.CollectArticleBean;
import com.hjc.reader.databinding.ItemArticleImageBinding;
import com.hjc.reader.databinding.ItemArticleTextBinding;

import org.jetbrains.annotations.NotNull;

import java.util.List;


public class ArticleAdapter extends BaseDelegateMultiAdapter<CollectArticleBean.DataBean.DatasBean, BaseViewHolder> {
    private static final int TYPE_TEXT = 1;
    private static final int TYPE_IMAGE = 2;

    public ArticleAdapter() {
        super();
        setMultiTypeDelegate(new CollectMultiTypeDelegate());
    }

    @Override
    protected void onItemViewHolderCreated(@NotNull BaseViewHolder viewHolder, int viewType) {
        DataBindingUtil.bind(viewHolder.itemView);
    }

    @Override
    protected void convert(@NotNull BaseViewHolder helper, CollectArticleBean.DataBean.DatasBean item) {
        if (item == null) {
            return;
        }
        switch (helper.getItemViewType()) {
            case TYPE_TEXT:
                initType1(helper, item);
                break;

            case TYPE_IMAGE:
                initType2(helper, item);
                break;
        }
    }

    private void initType1(BaseViewHolder helper, CollectArticleBean.DataBean.DatasBean item) {
        ItemArticleTextBinding binding = helper.getBinding();
        if (binding != null) {
            binding.setCollectBean(item);
        }
    }

    private void initType2(BaseViewHolder helper, CollectArticleBean.DataBean.DatasBean item) {
        ItemArticleImageBinding binding = helper.getBinding();
        if (binding != null) {
            binding.setCollectBean(item);
        }
    }

    // 代理类
    final static class CollectMultiTypeDelegate extends BaseMultiTypeDelegate<CollectArticleBean.DataBean.DatasBean> {

        public CollectMultiTypeDelegate() {
            // 绑定 item 类型
            addItemType(TYPE_TEXT, R.layout.item_article_text);
            addItemType(TYPE_IMAGE, R.layout.item_article_image);
        }

        @Override
        public int getItemType(@NotNull List<? extends CollectArticleBean.DataBean.DatasBean> data, int position) {
            CollectArticleBean.DataBean.DatasBean bean = data.get(position);
            String envelopePic = bean.getEnvelopePic();
            if (StringUtils.isEmpty(envelopePic)) {
                return TYPE_TEXT;
            } else {
                return TYPE_IMAGE;
            }
        }
    }
}

File: module-main/src/main/java/com/hjc/learn/main/adapter/ArticleAdapter.kt
package com.hjc.learn.main.adapter

import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import com.blankj.utilcode.util.StringUtils
import com.chad.library.adapter.base.BaseDelegateMultiAdapter
import com.chad.library.adapter.base.delegate.BaseMultiTypeDelegate
import com.chad.library.adapter.base.viewholder.BaseViewHolder
import com.hjc.learn.main.R
import com.hjc.learn.main.databinding.MainItemArticleImageBinding
import com.hjc.learn.main.databinding.MainItemArticleTextBinding
import com.hjc.library_net.model.WanCollectArticleBean

/**
 * @Author: HJC
 * @Date: 2021/3/3 10:28
 * @Description: 收藏文章adapter
 */
class ArticleAdapter : BaseDelegateMultiAdapter<WanCollectArticleBean, BaseViewHolder>() {

    companion object {
        private const val TYPE_TEXT = 1
        private const val TYPE_IMAGE = 2
    }

    init {
        setMultiTypeDelegate(CollectMultiTypeDelegate())
    }

    override fun onItemViewHolderCreated(viewHolder: BaseViewHolder, viewType: Int) {
        DataBindingUtil.bind<ViewDataBinding>(viewHolder.itemView)
    }

    override fun convert(holder: BaseViewHolder, item: WanCollectArticleBean) {
        when (holder.itemViewType) {
            TYPE_TEXT -> initType1(holder, item)
            TYPE_IMAGE -> initType2(holder, item)
        }
    }

    /**
     * 文字布局
     */
    private fun initType1(holder: BaseViewHolder, item: WanCollectArticleBean) {
        val binding = DataBindingUtil.getBinding<MainItemArticleTextBinding>(holder.itemView)
        binding?.collectBean = item
    }

    /**
     * 图片布局
     */
    private fun initType2(holder: BaseViewHolder, item: WanCollectArticleBean) {
        val binding = DataBindingUtil.getBinding<MainItemArticleImageBinding>(holder.itemView)
        binding?.collectBean = item
    }

    // 代理类
    class CollectMultiTypeDelegate : BaseMultiTypeDelegate<WanCollectArticleBean>() {

        init {
            // 绑定 item 类型
            addItemType(TYPE_TEXT, R.layout.main_item_article_text)
            addItemType(TYPE_IMAGE, R.layout.main_item_article_image)
        }

        override fun getItemType(data: List<WanCollectArticleBean>, position: Int): Int {
            val bean = data[position]
            val envelopePic: String? = bean.envelopePic

            return if (StringUtils.isEmpty(envelopePic)) {
                TYPE_TEXT
            } else {
                TYPE_IMAGE
            }
        }

    }

}